File: cpp/foxford_2nd_year/15/4_foxford.cpp
#include <iostream>
#include <vector>
#include <stack>
using namespace std;

int main()
{
    int n, k, p;
    int ans = 0;
    int count = 0;
    cin >> n >> k >> p;
    vector<stack<int>> q(k + 1);
    for (int i = 0; i < n; ++i)
    {
        char op;
        int a, b;
        cin >> op >> a >> b;
        if (op == '+')
        {
            q[a].push(b);
            ++count;
            if (count > p)
            {
                cout << "Error" << endl;
                return 0;
            }
            else
                ans = max(ans, count);
        }
        else
        {
            if (q[a].empty() || q[a].top() != b)
            {
                cout << "Error" << endl;
                return 0;
            }
            else
            {
                q[a].pop();
                --count;
            }
        }
    }
    if (count != 0)
        cout << "Error" << endl;
    else
        cout << ans << endl;
}

File: cpp/foxford_2nd_year/15/4_foxford.py
n, k, p = map(int, input().split())
 L = [[] for i in range(k + 1)]
 f = False
 ans = 0
 max_ans = 0
 for i in range(n):
     operation, otsek, toplivo = input().split()
     otsek, toplivo = int(otsek), int(toplivo)
     if operation == '+':
         if ans == p:
             f = True
         L[otsek].append(toplivo)
         ans += 1
         if max_ans < ans:
             max_ans = ans
     else:
         if L[otsek] == []:
             f = True
         elif L[otsek][-1] != toplivo:
             f = True
         else:
             L[otsek].pop()
             ans -= 1
 if ans != 0:
     f = True
 print('Error' if f else max_ans)